pkg/cas/content_addressable_storage_server.go

```diff
@@ -47,7 +47,36 @@ func (s *contentAddressableStorageServer) FindMissingBlobs(ctx context.Context,
 }
 
 func (s *contentAddressableStorageServer) BatchReadBlobs(ctx context.Context, in *remoteexecution.BatchReadBlobsRequest) (*remoteexecution.BatchReadBlobsResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "This service does not support batched reading of blobs")
+	// Asynchronously call Get() for every blob.
+	responsesChan := make(chan *remoteexecution.BatchReadBlobsResponse_Response, len(in.Digests))
+	for _, reqDigest := range in.Digests {
+		go func(reqDigest *remoteexecution.Digest) {
+			digest, err := digest.NewDigestFromPartialDigest(in.InstanceName, reqDigest)
+			var buf buffer.Buffer
+			if err == nil {
+				buf = s.contentAddressableStorage.Get(
+					ctx,
+					digest)
+			}
+			maxSizeBytes, err := buf.GetSizeBytes()
+			var data []byte
+			if err == nil {
+				data, err = buf.ToByteSlice(int(maxSizeBytes))
```

Comment:
This effectively puts no limit on the amount of memory a call to `BatchReadBlobs()` may consume. If I called `BatchReadBlobs()` with a thousand digests of 1 GB objects, this function would consume 1 TB of memory.

Considering that this is about limiting Protobuf message sizes, you could extend this type to also respect the `maximumMessageSizeBytes` configuration option. Some loop at the beginning of this function should be added to ensure the combined size of all blobs is not larger than `maximumMessageSizeBytes`.

Reply:
Would you be happy with `"Attempted to read %d bytes, while a maximum of %d bytes is permitted"` as the error message when exceeding maximumMessageSizeBytes?

Reply:
"Attempted to read a total of %d bytes, ..." maybe, to make it more explicit that it's the aggregate that counts?